cloud integration guides (Azure, Amazon) (#6403)
DevExpress/devextreme-documentation: concepts/Common/Integration Guides/70 Manage Azure Blob Storage with DevExtreme Components/30 FileManager.md

Full code examples:

* [Client-side data handling](https://github.com/DevExpress-Examples/devextreme-file-manager-azure-client-side-binding)
* [Server-side data handling](https://github.com/DevExpress-Examples/devextreme-file-manager-azure-server-side-binding)

> Azure Blob Storage doesn't expose a traditional file system to the end user. When you request to view the list of blob entries, Azure returns XML: 

Reviewer: ```suggestion
Azure Blob Storage does not expose a traditional file system to the end user. When you request to view the list of blob entries, Azure returns XML: 
```

    <EnumerationResults ServiceEndpoint="https://aspxdemos.blob.core.windows.net/" ContainerName="testfilemanager">
        <Blobs>
            <Blob>
                <Name>1.jpg</Name>
                <Properties>
                    <Creation-Time>Mon, 04 Dec 2023 07:24:58 GMT</Creation-Time>
                    <Last-Modified>Wed, 06 Dec 2023 06:29:09 GMT</Last-Modified>
                    <Etag>0x8DBF624A7904382</Etag>
                    <Content-Length>3126</Content-Length>
                    <Content-Type>application/octet-stream</Content-Type>
                    <Content-Encoding />
                    <Content-Language />
                    <Content-CRC64 />
                    <Content-MD5 />
                    <Cache-Control />
                    <Content-Disposition />
                    <BlobType>BlockBlob</BlobType>
                    <LeaseStatus>unlocked</LeaseStatus>
                    <LeaseState>available</LeaseState>
                    <ServerEncrypted>true</ServerEncrypted>
                </Properties>
                <OrMetadata />
            </Blob>
    ... (other Blob records)
        </Blobs>
        <NextMarker />
    </EnumerationResults>

> The `FileManager` component cannot work with such data. We need to perform additional actions to generate a FileManager-compatible filesystem representation.

Reviewer: `We need to perform additional actions to generate a FileManager-compatible filesystem representation.`

1. Although I do understand your explanation in the Trello card that this topic is a little "different" from the others, I'd like to rephrase this sentence without using the first person plural form "We" if possible.  How about "You", like we use in the rest of our documentation?

2. Is "actions" the best word to use here, or do you mean something like "steps" (if you're referring to the steps that follow in this topic)?

3. Is "filesystem" spelled correctly, or should it be "file system"?

> We can perform this task on the client or on the server.

Reviewer: We should update this line without using "perform". Do you mean "run this task"?

### Client-side Data Handling

> First, we'll create an API endpoint that handles FileManager requests and directs them to the Blob Storage:

Reviewer: ```suggestion
First, create an API endpoint that handles FileManager requests and directs them to the Blob Storage:
```

    [Route("api/file-manager-azure-access", Name = "FileManagerAzureAccessApi")]
    public object Process(string command, string blobName = "", string blobName2 = "")
    {
        try
        {
            return ProcessCommand(command, blobName, blobName2);
        }
        catch
        {
            return CreateErrorResult();
        }
    }
    object ProcessCommand(string command, string blobName, string blobName2)
    {
        switch (command)
        {
            case "BlobList":
                return GetBlobList();
            case "CreateDirectory":
                if (!AllowCreate)
                    return CreateErrorResult();
                return CreateDirectory(blobName);
                ...
        }
    }

> The `ProcessCommand` object defines the intenral logic of the endpoint. It calls functions that interact with Blob entities:

Reviewer: ```suggestion
The `ProcessCommand` object defines the internal logic of the endpoint. It calls functions that interact with Blob entities:
```

                isDirectory: false,
                dateModified: entry.lastModified,
                size: entry.length,
            };
            result.push(obj);
            }
        } else {
            const dirName = parts[0];
            let directory = directories[dirName];
            if (!directory) {
            directory = {
                name: dirName,
                isDirectory: true,
            };
            directories[dirName] = directory;
            result.push(directory);
            }
            if (!directory.hasSubDirectories) {
            directory.hasSubDirectories = parts.length > 2;
            }
        }
        });
        result.sort(this.compareDataObjects);
        return result;
    }

It outputs FileManager-compatible JSON:

    [
        {
            "name": "folder12",
            "isDirectory": true,
            "hasSubDirectories": true
        },
        ... other records
        {
            "name": "testflowers_137KB.jpgd1578202-9d72-4806-9ac2-90551d598e70",
            "isDirectory": false,
            "dateModified": "2022-12-21T14:57:55.000Z",
            "size": 140998
        }
    ]

### Server-side Data Handling

> If we want to process Blob data on the server, we need to create a class that implements the interface requirements of the FileManager component:

Reviewer: ```suggestion
To process Blob data on the server, create a class that implements the interface requirements of the FileManager component...
```

    public class AzureBlobFileProvider : IFileSystemItemLoader, IFileSystemItemEditor, IFileUploader, IFileContentProvider {
        ...
        public AzureBlobFileProvider(string storageAccountName, string storageAccessKey, string containerName, string tempDirPath) {
            ...
        }
    ...
    }

> And create function implementations for the Azure API:

Reviewer: ```suggestion
...and create function implementations for the Azure API:
```

    public IEnumerable<FileSystemItem> GetItems(FileSystemLoadItemOptions options) {
        var result = new List<FileSystemItem>();
        string dirKey = GetFileItemPath(options.Directory);
        var oneLevelItemsList = GetOneLevelHierarchyBlobs(dirKey);
        foreach(BlobHierarchyItem hierarchyItem in oneLevelItemsList) {
            var fileItem = GetFileSystemItem(hierarchyItem);
            if(fileItem != null) {
                result.Add(fileItem);
            }
        }
        return result.OrderByDescending(item => item.IsDirectory)
            .ThenBy(item => item.Name)
            .ToList();
    }

> The `FileSystem` end point will expose an instance of the `AzureBlobFileProvider` class to our client-side application:

Reviewer: ```suggestion
The `FileSystem` endpoint will expose an instance of the `AzureBlobFileProvider` class to the client-side application:
```

    [Route("api/file-manager-azure", Name = "FileManagerAzureProviderApi")]
    public object FileSystem(FileSystemCommand command, string arguments) {
        var config = new FileSystemConfiguration {
            Request = Request,
            FileSystemProvider = AzureFileProvider,
            ...
            UploadConfiguration = new UploadConfiguration {
                MaxFileSize = 1048576
            },
            TempDirectory = WebHostEnvironment.ContentRootPath + "/UploadTemp"
        };
        var processor = new FileSystemCommandProcessor(config);
        var result = processor.Execute(command, arguments);
        return result.GetClientCommandResult();
    }

> This approach greatly simplifies component setup. We can create a `RemoteFileSystemProvider` that uses the newly created filesystem endpoint:

Reviewer: 1. Please change "approach" to "technique".

2. Change the first person plural "We" to "You".

3. See my comment above regarding the spelling of "filesystem".

    const provider = new DevExpress.fileManagement.RemoteFileSystemProvider({
    endpointUrl: `${baseUrl}file-manager-azure`,
    });

> The `fileSystemProvider` component option simply needs to reference this provider:

Reviewer: ```suggestion
Note that the `fileSystemProvider` component option needs to reference the following provider:
```
